File: UltimateTicTacToe/tic_tac_toe.py

class TicTacToe:
    CROSS = 1
    NOUGHT = -1
    EMPTY = 0

    def __init__(self, rows=3, cols=3, win_set=3):

        if not isinstance(rows, TicTacToe):

            if rows <= 0 or cols <= 0 or win_set <= 0:
                raise Exception("rows, cols and win_set must be greater than zero.")
            if win_set > max(rows, cols):
                raise Exception("win_set must be less than rows and cols.")

            self.rows = rows
            self.cols = cols
            self.win_set = win_set
            self.moves_count = 0
            self.board = [([self.EMPTY] * self.cols) for _ in range(self.rows)]
            self.winner = None
        else:
            clone = rows
            self.rows = clone.rows
            self.cols = clone.cols
            self.win_set = clone.win_set
            self.moves_count = clone.moves_count
            self.board = [([self.EMPTY] * self.cols) for _ in range(self.rows)]
            self.winner = clone.winner

            for row in range(self.rows):
                for col in range(self.cols):
                    self.board[row][col] = clone.board[row][col]

    def check_winner(self):
        position = self.turn_position()
        draw = True

        for row in range(self.rows):
            for col in range(self.cols):
                if self.board[row][col] == position:
                    draw = False
                    for i in range(-1, 2):
                        for j in range(-1, 2):
                            if i is j:
                                continue
                            for w in range(self.win_set):
                                if not self.in_bounds(self.rows + i, self.cols + j) or self.board[self.rows + i][self.cols + j] != position:
                                    break
                            else:
                                self.winner = position
                                return
        if draw:
            self.winner = self.EMPTY

    def set(self, row, col, position):
        if self.in_bounds(row, col) and position != self.EMPTY and self.board[row][col] == self.EMPTY:
            self.board[row][col] = position
            self.check_winner()
            self.moves_count += 1
            return True
        return False

    def play(self, row, col):
        position = self.turn_position()
        return self.set(row, col, position)

    def turn_position(self):
        return self.CROSS if self.moves_count % 2 == 0 else self.NOUGHT

    def in_bounds(self, row, col):
        return 0 <= row < self.rows and 0 <= col < self.cols
    
    def game_over(self):
        return self.winner is not None

    def clone(self):
        return TicTacToe(self)

File: UltimateTicTacToeCS/TicTacToe.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace UltimateTicTacToeCS
{
    public class TicTacToe
    {
        public const int ROWS = 3;
        public const int COLS = 3;
        public const int WIN_SET = 3;

        public static SqrState TurnToSqrState(Turn turn)
        {
            if (turn == Turn.Cross)
            {
                return SqrState.Cross;
            }

            return SqrState.Nought;
        }

        public enum SqrState { Nought = -1, Empty, Cross }
        public enum Turn { Cross, Nought }
        public enum WinState { Nought=-1, Draw, Cross, NoOne}

        public SqrState[,] Board { get; private set; }
        public int Moves { get; set; }
        public Turn GameTurn { get; set; }
        public WinState Winner { get; private set; }
        public bool GameOver => Winner != WinState.NoOne;
        public TicTacToe Clone => new TicTacToe(this);

        public TicTacToe(TicTacToe clone)
        {
            Board = new SqrState[ROWS, COLS];
            Moves = clone.Moves;
            GameTurn = clone.GameTurn;
            Winner = clone.Winner;

            for (int row = 0; row < ROWS; ++row)
            {
                for (int col = 0; col < COLS; ++col)
                {
                    Board[row, col] = clone.Board[row, col];
                }
            }
        }

        public TicTacToe()
        {
            NewBoard();
        }

        public void NewBoard()
        {
            Board = new SqrState[ROWS, COLS];
            Moves = 0;
            Winner = WinState.NoOne;
            GameTurn = Turn.Cross;
        }

        public bool Play(int row, int col)
        {
            if (InBounds(row, col) && Board[row, col] == SqrState.Empty && !GameOver)
            {
                Board[row, col] = TurnToSqrState(GameTurn);
                CheckWin();
                ++Moves;
                NextTurn();
                return true;
            }

            return false;
        }

        public void NextTurn()
        {
            GameTurn = GameTurn == Turn.Cross ? Turn.Nought : Turn.Cross;
        }

        public void CheckWin()
        {
            var state = TurnToSqrState(GameTurn);
            bool haveMoves = false;

            for (int row = 0; row < ROWS; ++row)
            {
                for (int col = 0; col < COLS; ++col)
                {
                    if (Board[row, col] == state)
                    {
                        for (int dRow = -1; dRow < 2; ++dRow)
                        {
                            for (int dCol = -1; dCol < 2; ++dCol)
                            {
                                if (dRow == 0 && dCol == 0)
                                {
                                    continue;
                                }

                                int w = 1;
                                for (; w < WIN_SET; w++)
                                {
                                    if (!InBounds(row + dRow * w, col + dCol * w) || Board[row + dRow * w, col + dCol * w] != state)
                                    {
                                        break;
                                    }
                                }

                                if (w == WIN_SET)
                                {
                                    Winner = (WinState)state;
                                    return;
                                }
                            }
                        }
                    }
                    else if (Board[row, col] == SqrState.Empty)
                    {
                        haveMoves = true;
                    }
                }
            }

            if (!haveMoves)
            {
                Winner = WinState.Draw;
            }
        }

        public bool InBounds(int row, int col)
        {
            return Math.Max(0, Math.Min(row, ROWS - 1)) == row &&
                Math.Max(0, Math.Min(col, COLS - 1)) == col;
        }
    }
}
